crates/pico/tests/memoization.rs

```diff
@@ -0,0 +1,93 @@
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        LazyLock, Mutex,
+    },
+};
+
+use calc::{ast::Program, error::Result, eval::eval, lexer::Lexer, parser::Parser};
+use pico::storage::DefaultStorage;
+use pico_core::{database::Database, source::SourceId, storage::Storage};
+use pico_macros::{memo, Db, Source};
+
+mod calc;
+
+static SUM_COUNTER: AtomicUsize = AtomicUsize::new(0);
+static EVAL_COUNTER: LazyLock<Mutex<HashMap<SourceId<Input>, usize>>> =
+    LazyLock::new(|| Mutex::new(HashMap::new()));
+
+#[test]
+fn memoization() {
+    let mut state = TestDatabase::default();
+
+    let left = state.set(Input {
+        key: "left",
+        value: "2 + 2 * 2".to_string(),
+    });
+
+    let right = state.set(Input {
+        key: "right",
+        value: "(2 + 2) * 2".to_string(),
+    });
+
+    let result = sum(&mut state, left, right);
+    assert_eq!(result, 14);
+    assert_eq!(state.storage().current_epoch(), 3.into());
```

Comment:
though I understand why `.current_epoch()` needs to be exposed on the `trait Storage` (because it's used by `trait Db`), it feels to me like an implementation detail from the perspective of the developer. Perhaps we should avoid making assertions about `current_epoch` here?